File: reverse-engineering/modbus-command.js
class ModbusCommand {
  constructor () {
    this.CRC16H = 0
    this.CRC16L = 0

    this.address = {
      coil: 4097,
      disk: 8193,
      hold: 12289,
      input: 16385
    }

    /*
    this.client_buffer_size = 7
    this.client_coil_size = 47
    this.client_disk_size = 5
    this.client_hold_size = 52
    this.client_input_size = 31
    this.comm_max_failed_times = 4
    this.delay_time_setup = 10
    this.error_addr = 3
    this.error_crc = 1
    this.error_func = 2
    this.error_operate = 5
    this.error_time_out = 6
    this.error_value = 4
    this.size_coil = 20
    this.size_disk = 20
    this.size_hold = 80
    this.size_input = 20
    this.success = 0
    this.time_out_setup = 3000
    this.net_connect_failed = 11
    this.net_connect_success = 10
    this.net_connect_tcpclose = 12
    this.ip = null
    this.commFaiedTimes = 0
    this.canWrite = true
    this.flag = false
    */

    this.valueCoil = new Array(20).fill(0)
    this.valueDisc = new Array(20).fill(0)
    this.valueHold = new Array(80).fill(0)
    this.valueInput = new Array(20).fill(0)
    this.list = []
  }

  /*
  getIp() {
    return this.ip
  }

  setIp(ip) {
    this.ip = ip
  }

  addCommFaiedTimes () {
    this.commFaiedTimes++
  }

  clearCommFaiedTimes () {
    this.commFaiedTimes = 0
  }

  isCommFailed () {
    return this.commFaiedTimes === 4
  }
  */

  addCommand (cmd) {
    // this.canWrite = false
    this.list.push(cmd)
  }

  clearCommand () {
    this.list = []
  }

  isEmpty () {
    return this.list.length < 1
  }

  removeFirst () {
    this.list.shift()
  }

  CRC16 (bufData, buflen) {
    let CRC = 65535
    if (buflen !== 0) {
      for (let i = 0; i < buflen; i++) {
        CRC ^= bufData[i] & 255
        for (let j = 0; j < 8; j++) {
          if ((CRC & 1) !== 0) {
            CRC = (CRC >> 1) ^ 40961
          } else {
            CRC >>= 1
          }
        }
      }
      this.CRC16L = ((CRC & 255) | 0)
      this.CRC16H = ((CRC >> 8) | 0)
    }
    return 0
  }
}

module.exports = new ModbusCommand()

File: reverse-engineering/decompiled/ModbusCommand.java
package com.kamoer.x1dosingpump.sockets;

import android.util.Log;
import java.io.Serializable;
import java.util.ArrayList;
import java.util.List;

public class ModbusCommand implements Serializable {
    public static byte CRC16H = 0;
    public static byte CRC16L = 0;
    public static final int MODBUS_ADDR_COIL = 4097;
    public static final int MODBUS_ADDR_DISK = 8193;
    public static final int MODBUS_ADDR_HOLD = 12289;
    public static final int MODBUS_ADDR_INPUT = 16385;
    public static final int MODBUS_CLIENT_BUFFER_SIZE = 7;
    public static final int MODBUS_CLIENT_COIL_SIZE = 47;
    public static final int MODBUS_CLIENT_DISK_SIZE = 5;
    public static final int MODBUS_CLIENT_HOLD_SIZE = 52;
    public static final int MODBUS_CLIENT_INPUT_SIZE = 31;
    public static final int MODBUS_COMM_MAX_FAILED_TIMES = 4;
    public static final int MODBUS_DELAY_TIME_SETUP = 10;
    public static final int MODBUS_ERROR_ADDR = 3;
    public static final int MODBUS_ERROR_CRC = 1;
    public static final int MODBUS_ERROR_FUNC = 2;
    public static final int MODBUS_ERROR_OPERATE = 5;
    public static final int MODBUS_ERROR_TIME_OUT = 6;
    public static final int MODBUS_ERROR_VALUE = 4;
    public static final int MODBUS_SIZE_COIL = 20;
    public static final int MODBUS_SIZE_DISK = 20;
    public static final int MODBUS_SIZE_HOLD = 80;
    public static final int MODBUS_SIZE_INPUT = 20;
    public static final int MODBUS_SUCCESS = 0;
    public static final int MODBUS_TIME_OUT_SETUP = 3000;
    public static final int NET_CONNECT_FAILED = 11;
    public static final int NET_CONNECT_SUCCESS = 10;
    public static final int NET_CONNECT_TCPCLOSE = 12;
    private static ModbusCommand command = null;
    public boolean canWrite;
    public List<String> cmdList;
    private int commFaiedTimes;
    public boolean flag;
    public String ip;
    public int[] valueCoil;
    public int[] valueDisc;
    public int[] valueHold;
    public int[] valueInput;

    public ModbusCommand() {
        this.valueCoil = null;
        this.valueDisc = null;
        this.valueInput = null;
        this.valueHold = null;
        this.cmdList = null;
        this.commFaiedTimes = 0;
        this.canWrite = true;
        this.flag = false;
        this.valueCoil = new int[20];
        this.valueDisc = new int[20];
        this.valueHold = new int[80];
        this.valueInput = new int[20];
        this.cmdList = new ArrayList();
    }

    public static ModbusCommand getInstance() {
        if (command == null) {
            command = new ModbusCommand();
        }
        return command;
    }

    public String getIp() {
        return this.ip;
    }

    public void setIp(String ip2) {
        this.ip = ip2;
    }

    public void addCommand(String cmd) {
        Log.i("rock-cmd", cmd);
        this.canWrite = false;
        this.cmdList.add(cmd);
    }

    public void clearCommand() {
        this.cmdList.clear();
    }

    public boolean isEmpty() {
        return this.cmdList.isEmpty();
    }

    public void removeFirst() {
        this.cmdList.remove(0);
    }

    public void addCommFaiedTimes() {
        this.commFaiedTimes++;
    }

    public void clearCommFaiedTimes() {
        this.commFaiedTimes = 0;
    }

    public boolean isCommFailed() {
        return this.commFaiedTimes == 4;
    }

    public static int Funct_CRC16(byte[] bufData, int buflen) {
        int CRC = 65535;
        if (buflen != 0) {
            for (int i = 0; i < buflen; i++) {
                CRC ^= bufData[i] & 255;
                for (int j = 0; j < 8; j++) {
                    if ((CRC & 1) != 0) {
                        CRC = (CRC >> 1) ^ 40961;
                    } else {
                        CRC >>= 1;
                    }
                }
            }
            CRC16L = (byte) (CRC & 255);
            CRC16H = (byte) (CRC >> 8);
        }
        return 0;
    }
}
